README.md
# NuvlaEdge Agent

[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg?style=for-the-badge)](https://github.com/nuvlaedge/agent/graphs/commit-activity)
[![GitHub issues](https://img.shields.io/github/issues/nuvlaedge/agent?style=for-the-badge&logo=github&logoColor=white)](https://GitHub.com/nuvlaedge/agent/issues/)
[![Docker pulls](https://img.shields.io/docker/pulls/nuvlaedge/agent?style=for-the-badge&logo=Docker&logoColor=white)](https://cloud.docker.com/u/nuvlaedge/repository/docker/nuvlaedge/agent)
[![Docker image size](https://img.shields.io/docker/image-size/nuvladev/agent/master?logo=docker&logoColor=white&style=for-the-badge)](https://cloud.docker.com/u/nuvlaedge/repository/docker/nuvlaedge/agent)

![CI Build](https://github.com/nuvlaedge/agent/actions/workflows/main.yml/badge.svg)
![CI Release](https://github.com/nuvlaedge/agent/actions/workflows/release.yml/badge.svg)


**This repository contains the source code for the NuvlaEdge Agent - the microservice which is responsible for the [NuvlaEdge](https://sixsq.com/products-and-services/nuvlaedge/overview) activation, categorization and telemetry.**

This microservice is an integral component of the NuvlaEdge Engine.

---

**NOTE:** this microservice is part of a loosely coupled architecture, thus when deployed by itself, it might not provide all of its functionalities. Please refer to https://github.com/nuvlaedge/deployment for a fully functional deployment

---

## Build the NuvlaEdge Agent

This repository is already linked with Travis CI, so with every commit, a new Docker image is released.

There is a [POM file](pom.xml) which is responsible for handling the multi-architecture and stage-specific builds.

**If you're developing and testing locally in your own machine**, simply run `docker build .` or even deploy the microservice via the local [compose files](docker-compose.yml) to have your changes built into a new Docker image, and saved into your local filesystem.

**If you're developing in a non-master branch**, please push your changes to the respective branch, and wait for Travis CI to finish the automated build. You'll find your Docker image in the [nuvladev](https://hub.docker.com/u/nuvladev) organization in Docker hub, names as _nuvladev/agent:\<branch\>_.

## Deploy the NuvlaEdge Agent

The NuvlaEdge agent will only work if a [Nuvla](https://github.com/nuvla/deployment) endpoint is provided and a NuvlaEdge has been added in Nuvla.

### Prerequisites

 - *Docker (version 18 or higher)*
 - *Docker Compose (version 1.23.2 or higher)*

### Environment variables

|                          	|                                                                                                                                                       	|
|-------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------	|
|           NUVLAEDGE_UUID 	| (**required**) before starting the NuvlaEdge Agent, make sure you export the ID of the NuvlaEdge you've created through Nuvla: `export NUVLAEDGE_UUID=<nuvlaedge id from nuvla>` 	|
| NUVLA_ENDPOINT_INSECURE 	| if you're using an insecure Nuvla endpoint, set this to `True`: `export NUVLA_ENDPOINT_INSECURE=True`                                                	|
|          NUVLA_ENDPOINT 	| if you're not using [nuvla.io](https://nuvla.io) then set this to your Nuvla endpoint: `export NUVLA_ENDPOINT=<your endpoint>`                                      	|
| | |

### Launching the NuvlaEdge Agent

Simply run `docker-compose up --build`

## Test the NuvlaEdge Agent

This microservice is completely automated, meaning that as long as all the proper environment variables have been correctly set and the right dependencies have been met, the respective Docker container will start by itself, automatically activate the NuvlaEdge in Nuvla and start sending telemetry to periodically.

### Agent API Calls

The NuvlaEdge Agent provides an internal REST API for allowing other NuvlaEdge component to interact with Nuvla without having to implement their own Nuvla API clients, thus reducing duplicated code throughout the NuvlaEdge software stack.

**The API is not published outside the Docker network, and is available at _http://agent/api_**

#### Get agent healthcheck

Returns something if the agent and respective API are already up and running. This call is meant to be used as a healthcheck for other components that are waiting for the agent to be ready.

 - **URL**

    /api/healthcheck

 - **Methods**

    `GET`

 - **URL Params**

    None

 - **On success**
    - Code: `200`

        Content: `True`

 - **Example**

    ```
    curl --fail -X GET http://agent/api/healthcheck
    ```   

#### Re-commission the NuvlaEdge

It might happen that due to a change in the NuvlaEdge or a need to refresh certain credentials and/or system configurations, the NuvlaEdge needs to be re-commissioned so that Nuvla can propagate the new changes. This call will trigger a NuvlaEdge re-commissioning, according to the request payload.

This call is tailored to be used by specific NuvlaEdge components like the Network Manager.


#### Manage NuvlaEdge peripherals

Gives other NuvlaEdge components the ability to register and manage NuvlaEdge peripherals in Nuvla.

##### Register a new NuvlaEdge peripheral

Creates a new nuvlabox-peripheral resource in Nuvla and save a local copy of this peripheral locally in the NuvlaEdge.


 - **URL**

    /api/peripheral

 - **Methods**

    `POST`

 - **URL Params**

    None

  - **Data payload**

    JSON document with a structure matching the `nuvlabox-peripheral` [resource in Nuvla](https://nuvla.io/ui/documentation/nuvlabox-peripheral-1-1)

 - **On success**
    - Code: `201`

        Content: JSON response from Nuvla. Example:

                ```
                {
                    "status": 201,
                    "message": "some message",
                    "resource-id": "<uuid of the Nuvla resource>"
                }
                ```

 - **On error**
    - Code: `400`

        Content: `{"error": "error message"}`

    - Code: `500`

        Content: `{"error": "error message"}`

    - Code: others

        Content: If the error has been thrown during the interaction with the Nuvla API, then the Nuvla response will be literally forwarded. This response might include error codes like `403`, `404`, `409`, etc.


 - **Example**

    ```
    curl --fail -X POST http://agent/api/peripheral \
            -H content-type:application/json \
            -d '''{
                    "available": true,
                    "classes": ["audio", "video"],
                    "name": "my peripheral",
                    "description": "description of my awesome peripheral",
                    "device-path": "/dev/mydevice",
                    "identifier": "unique-local-peripheral-id",
                    "interface": "USB",
                    "product": "Gadget A",
                    "serial-number": "1234567890",
                    "vendor": "company A"
                   }'''

    ```   

    Response:
    ```
    {
        "message":"nuvlabox-peripheral/d149dc67-fd5c-402d-8608-306852a8c0f3 created",
        "resource-id":"nuvlabox-peripheral/d149dc67-fd5c-402d-8608-306852a8c0f3",
        "status":201    
    }
    ```

##### Search for a NuvlaEdge peripheral

Returns a list of NuvlaEdge peripherals matching the URL params in the request.

 - **URL**

    /api/peripheral

 - **Methods**

    `GET`

 - **URL Params**

    `parameter=<peripheral-parameter-name>`

    `value=<peripheral-parameter-value>`

    `identifier_pattern=<pattern-for-multiple-peripheral-identifiers>`


  - **Data payload**

    None

 - **On success**
    - Code: `200`

        Content: Dict of matching peripheral identifiers

 - **Examples**

    ```
    curl --fail -X GET 'http://agent/api/peripheral?parameter=interface&value=USB'
    ```   

    Response:
    ```
    {"peripheral-identifier": {"interface": ...}}
    ```
    ---


    ```
    curl --fail -X GET 'http://agent/api/peripheral?parameter=interface&value=BT'
    ```   

    Response:
    ```
    {}
    ```

    ---


    ```
    curl --fail -X GET 'http://agent/api/peripheral?identifier_pattern=unique-local*'
    ```   

    Response:
    ```
    {"peripheral-1-id": {...}, "peripheral-2-id": {...}}
    ```


##### Manage an existing peripheral

Provides reading and editing capabilities for an existing peripheral.

 - **URL**

    /api/peripheral/<identifier>

 - **Methods**

    `DELETE`

 - **URL Params**

    None

  - **Data payload**

    None

 - **On success**
    - Code: `200`

        Content: `{"message": "successfully deleted ..."}` (can contain extra fields like `status` and `resource-id`)

 - **On error**
    - Code: `404`

        Content: `{"error": "not found error message"}`

    - Code: `500`

        Content: `{"error": "error message"}`

    - Code: others

        Content: If the error has been thrown during the interaction with the Nuvla API, then the Nuvla response will be literally forwarded. This response might include error codes like `403`, `404`, `409`, etc.


 - **Example**

    ```
    curl -X DELETE http://agent/api/peripheral/bad-local-peripheral
    ```   

    Response:
    ```
    {
        "error":"Peripheral not found"
    }
    ```



## Contributing

This is an open-source project, so all community contributions are more than welcome. Please read [CONTRIBUTING.md](CONTRIBUTING.md)

## Copyright

Copyright &copy; 2021, SixSq SA


CHANGELOG.md
# Changelog

## Unreleased
### Added
### Changed

## [2.12.0] - 2023-05-22
### Added
 - NuvlaEdge Docker client, now reads the port from the container configuration 
### Changed
 - Compute API Connectivity check only performed in local network using container-name and internal port
 - Removed tag push from commissioning process

## [2.11.0] - 2023-04-27
### Added
 - Integration of nuvlaedge-common library
 - PeripheralManager thread to consume from peripheral reports
 - PeripheralManagerDB handler 
 - Added nuvlaedge-common builder in the main.yml for testing
 - Split on main.yml into jobs for readability
### Changed
 - Peripherals no longer communicate with API rest but via file
 - File definitions are no longer class attributes but constants defined in nuvlaedge-common library
 - UnitTests adapted to new nuvlaedge-common library

## [2.10.0] - 2023-04-26
### Added
### Changed
 - Remove Agent API
 - Refactor Agent
 - Ensure every requests calls have a timeout
 - Recommission VPN if OpenVPN config is not available and do nothing is VPN client container doesn't exist
 - Do not wait on system-manager to Activate the NuvlaEdge
 - Explicit container naming based on project name
 - Bugfix of install_ssh_key and get_working_dir_from_labels
 - Improved test of Compute API
 - Job executor containers: copy environment variables and volumes (bind mounts) from the job-engine-lite container 
 - Improved and fixed retrieval of project name and NuvlaEdge containers.

## [2.9.6] - 2022-12-07
### Added
### Changed
 - Telemetry: update network/interfaces to a list of dict
 - (re)write files in an atomic way
 - NuvlaEdgeCommon: refactor set_nuvlaedge_id() and use credential alternatively

## [2.9.5] - 2022-11-16
### Added 
 -  Port of 2.8.5: Default gw report (#118) 
### Changed

## [2.9.4] - 2022-10-31
### Added
 - Print threads stacktraces on SIGUSR1 [27961e8]
### Changed
 - Vpn client: do not open port [83f8e5d]
 - Port of: 2.8 telemetry fixes (#126) [56765ff]

## [2.9.3] - 2022-09-26
### Added
### Changed
 - Fix bug where installation-parameters working-dir and config-files were not set

## [2.9.2] - 2022-09-22
### Added
### Changed
 - Port of fixes from 2.8.1 and 2.8.2

## [2.9.1] - 2022-09-05
### Added
 - Add code lost during Telemetry refactor (#114)
### Changed

## [2.9.0] - 2022-08-31
### Added
### Changed
 - Migration to NuvlaEdge

## [2.8.8] - 2023-01-25
### Added
### Changed
 - Ensure every requests calls have a timeout
 
## [2.8.7] - 2022-12-07
### Added
### Changed
 - (re)write files in an atomic way
 - NuvlaBoxCommon: refactor set_nuvlabox_id() and use credential alternatively
 
## [2.8.6] - 2022-12-05
### Added
### Changed
 - Telemetry: update network/interfaces to a list of dict

## [2.8.5] - 2022-11-16
### Added 
 - Telemetry: report all detected IPs and IPs of all network interfaces 
 - Telemetry: report default gateway interface
### Changed

##[2.8.4] - 2022-10-27
### Added
 - Added support for all threaded monitors
 - Added all monitors to thread as default. Added env_variable return them to normal exec
### Changed
 - Fix late start of monitor threads
 - Captured exception caused by late start of the compute-api in the Agent
 - Change IP route container to detach mode and manually deleted after retrieving the logs

## [2.8.3] - 2022-09-24
### Changed
 - Fix bug where installation-parameters working-dir and config-files were not set (backport)

## [2.8.2] - 2022-09-22
### Added
 - add code lost during refactor (cherry-pick)
### Changed
 - container_stats.py: fixed possible None value
 - fix tests

## [2.8.1] - 2022-09-07
### Added
 - Add missing dependency lsblk
 - Add timeout to requests to geolocation services
### Changed
 - Set missing default socket timeout

## [2.8.0] - 2022-07-18
### Added
 - Added new Pydantic BaseModel to validate data from Nuvla.
 - Added a modular monitor structure in which every monitor can be included or excluded.
   (Some are compulsory)
 - Added data structure for each monitor to allow data validation
 - New future data structure provided to IP interfaces. Still need to adapt
   server-side to enable it in Agent.
 - Added host device network interfaces reading. Using an auxiliary Docker
   container connected to host network.
 - Added public IP(v4&v6) reading via ping to external server.
### Changed
 - Use common base image for all NE components
 - Telemetry class now is a watchdog of monitors. It keeps them running and starts them
   over if needed
 - Whole telemetry class has been refactored
 - Logging formatting modified and unified
 - Logging main configuration parsed from config file
 - Change Dockerfile to match common python3.8-alpine3.12 NE engine versions
 - Removed Wiring Pi from docker. TODO: remove GPIO monitor too, maybe to be provided as a peripheral

## [2.7.2] - 2022-04-18
### Added
### Changed
 - Do not fail if some installation parameters are not found
 - Removed version pinning for Flask

## [2.7.1] - 2022-04-12
### Added
 - requirements.txt: added itsdangerous==2.0.1
### Changed

## [2.7.0] - 2022-04-11
### Added
 - New env var for vpn extra config
### Changed

## [2.6.0] - 2022-03-24
### Added
 - Add org.opencontainers labels
 - Make logging level configurable
### Changed
 - Check if provided UUID is different from old one - avoid overwriting
 - NuvlaBox log: add components to telemetry
 - Only take the config files from the last update

## [2.5.0] - 2021-12-16
### Added
 - Enable compression when sending data to api-server
### Changed
 - Code refactor and bugfixes

## [2.4.1] - 2021-11-29
### Added
### Changed
 - Improved telemetry to send all changes to Nuvla and reduce the chance of a "split brain" like scenario
 - Delete ignored files

## [2.4.0] - 2021-11-03
### Added
 - Kubernetes infrastructure discovery for kubeadm, k0s and k3s installations
### Changed
 - Improved parallelization of telemetry and heartbeat cycles
 - Fixed IP-based geolocation retrieval
 - General fixes and code refactoring

## [2.3.4] - 2021-10-19
### Added

code/old_app.py
#!/usr/local/bin/python
# -*- coding: utf-8 -*-

"""NuvlaBox Agent service

This service takes care of the NuvlaBox activation and subsequent
resource discovery and categorization within the hosting device.

It is also responsible for all telemetry data sent to Nuvla.

Arguments:
:param d/debug: (optional) log level set to DEBUG
:param q/quiet: (optional) log level set to CRITICAL
:param l/log_filepath: (optional) path to the log file
:param v/volume: (optional) shared volume where all NuvlaBox data can be found
"""

import argparse
import socket
import threading
import json
import logging
import sys
import os
import requests
import time
from flask import Flask, request, jsonify


from threading import Event
from agent.common import NuvlaBoxCommon
from agent.activate import Activate
from agent.telemetry import Telemetry
from agent.infrastructure import Infrastructure
from agent.job import Job

__copyright__ = "Copyright (C) 2019 SixSq"
__email__ = "support@sixsq.com"

root_logger: logging.Logger = logging.getLogger('agent')
log_format: str = '[%(asctime)s - %(name)s/%(funcName)s - %(levelname)s]: %(message)s'
formatter: logging.Formatter = logging.Formatter(log_format)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="NuvlaBox Agent")
    parser.add_argument('--debug', dest='debug', default=False, action='store_true',
                        help='use for increasing the verbosity level')

    if parser.parse_args().debug:
        logging.basicConfig(
            stream=sys.stdout,
            level=logging.DEBUG,
            format=log_format)
    else:
        try:
            logging.basicConfig(
                stream=sys.stdout,
                level=logging.getLevelName(os.getenv('NUVLABOX_LOG_LEVEL')),
                format=log_format)
        except ValueError:
            logging.basicConfig(
                stream=sys.stdout,
                level=logging.INFO,
                format=log_format)

# finish imports with logging
import agent.agent_api as AgentApi

app = Flask(__name__)
data_volume = "/srv/nuvlabox/shared"
default_log_filename = "agent.log"
network_timeout = 10


def manage_pull_jobs(job_list, job_image_name):
    """
    Handles the pull jobs one by one, sequentially

    :param job_list: list of job IDs
    :param job_image_name: Docker Image to be used for the job-engine
    """
    for job_id in job_list:
        job = Job(data_volume, job_id, job_image_name)
        if job.do_nothing:
            continue

        try:
            job.launch()
        except Exception as ex:
            # catch all
            root_logger.error(f'Cannot process job {job_id}. Reason: {str(ex)}')


def preflight_check(activation_class_object: Activate,
                    infra_class_object: Infrastructure,
                    nb_updated_date: str):
    """
    Checks if the NuvlaBox resource has been updated in Nuvla
    :param activation_class_object: instance of Activate
    :param infra_class_object: instance of Infrastructure
    :param nb_updated_date: date of the last NB resource update
    :return:
    """
    nuvlabox_resource = activation_class_object.get_nuvlabox_info()

    global refresh_interval
    global can_continue
    global nuvlabox_info_updated_date

    if nuvlabox_resource.get('state', '').startswith('DECOMMISSION'):
        root_logger.warning(f'This NuvlaBox is {nuvlabox_resource["state"]} in Nuvla. '
                            f'Exiting...')
        can_continue = False

    if nb_updated_date != nuvlabox_resource['updated'] and can_continue:
        refresh_interval = nuvlabox_resource['refresh-interval']
        root_logger.warning(f'NuvlaBox resource updated. Refresh interval value: '
                            f'{refresh_interval}s')
        nuvlabox_info_updated_date = nuvlabox_resource['updated']
        old_nuvlabox_resource = \
            activation_class_object.create_nb_document_file(nuvlabox_resource)
        activation_class_object.vpn_commission_if_needed(nuvlabox_resource,
                                                         old_nuvlabox_resource)

    # if there's a mention to the VPN server, then watch the VPN credential
    if nuvlabox_resource.get("vpn-server-id"):
        infra_class_object.watch_vpn_credential(nuvlabox_resource.get("vpn-server-id"))


@app.route('/api/status')
def set_status():
    """ API endpoint to let other components set the NuvlaBox status """

    value = request.args.get('value')
    log = str(request.args.get('log'))

    if not value:
        root_logger.warning("Received status request with no value. Nothing to do")
    else:
        root_logger.info("Setting NuvlaBox status to {}".format(value))
        if log:
            print(app.config["telemetry"], dir(app.config["telemetry"]))

    root_logger.warning('to be implemented')
    return "to be implemented"


@app.route('/api/find-data-gateway')
def find_data_gateway():
    """
    Returns 200 or 404, depending on whether the data-gateway is reachable or not

    :return: 200 or 404
    """

    try:
        socket.gethostbyname('data-gateway')
        return jsonify('success'), 200
    except socket.gaierror as e:
        return jsonify(str(e)), 404


@app.route('/api/commission', methods=['POST'])
def trigger_commission():
    """ API endpoint to let other components trigger a commissioning

    The request.data is the payload
    """

    payload = json.loads(request.data)

    root_logger.info('Commission triggered via the NB Agent API')

    commissioning_response = app.config["infra"].do_commission(payload)
    return jsonify(commissioning_response)


@app.route('/api/set-vulnerabilities', methods=['POST'])
def set_vulnerabilities():
    """ API endpoint to let other components send and save the scanned vulnerabilities

    The request.data is the payload
    """

    payload = json.loads(request.data)

    root_logger.info('Saving vulnerabilities received via the API: %s ' % payload)

    AgentApi.save_vulnerabilities(payload)

    return jsonify(True), 201


@app.route('/api/set-vpn-ip', methods=['POST'])
def set_vpn_ip():
    """ API endpoint to let other components define the NB VPN IP
    """

    payload = request.data

    root_logger.info('Received request to set VPN IP to %s ' % payload)

    AgentApi.save_vpn_ip(payload)
    return jsonify(True), 201


@app.route('/api/healthcheck', methods=['GET'])
def healthcheck():
    """ Static endpoint just for clients to check if API/Agent is up and running
    """

    return jsonify(True)


@app.route('/api/agent-container-id', methods=['GET'])
def get_agent_container_id():
    """ Static endpoint just for clients to get the Agent container Docker ID
    """

    return jsonify(socket.gethostname())


@app.route('/api/peripheral', defaults={'identifier': None}, methods=['POST', 'GET'])
@app.route('/api/peripheral/<path:identifier>', methods=['GET', 'PUT', 'DELETE'])
def manage_peripheral(identifier):
    """ API endpoint to let other components manage NuvlaBox peripherals

    :param identifier: local id of the peripheral to be managed
    """

    root_logger.info('  ####   Received %s request for peripheral management' % request.method)

    payload = {}
    if request.data:
        try:
            payload = json.loads(request.data)
        except:
            return jsonify({"error": "Payload {} malformed. It must be a JSON payload".format(payload)}), 400

    if identifier:
        root_logger.info('  ####   %s peripheral %s' % (request.method, identifier))
        if request.method in ["DELETE", "PUT"]:
            # DELETE accepts resource ID for simplicity and backward compatibility
            resource_id = request.args.get('id')
            message, return_code = AgentApi.modify(identifier, peripheral_nuvla_id=resource_id,
                                                   action=request.method, payload=payload)
        elif request.method == "GET":
            message, return_code = AgentApi.get(identifier)
        else:
            root_logger.info('  ####   Method %s not implemented yet!!' % request.method)
            message = "Not implemented"
            return_code = 501
    else:
        # POST or FIND peripheral
        if request.method == "POST":
            root_logger.info('  ####   Creating new peripheral with payload %s' % payload)
            message, return_code = AgentApi.post(payload)
        else:
            # GET
            parameter = request.args.get('parameter')
            value = request.args.get('value')
            identifier_pattern = request.args.get('identifier_pattern')
            root_logger.info('  ####   Find peripherals with {}={}'.format(parameter, value))
            message, return_code = AgentApi.find(parameter, value, identifier_pattern)

    return jsonify(message), return_code


def send_heartbeat(nb_instance, nb_telemetry, nb_status_id: str, previous_status_time: str):
    """
    Updates the NuvlaBox Status according to the local status file
    :param nb_instance: instance of class NuvlaBoxCommon.NuvlaBoxCommon()
    :param nb_telemetry: instance of class Telemetry()
    :param nb_status_id: ID of the NB status resource
    :param previous_status_time: ISO timestamp of the previous status heartbeat
    :return: (Nuvla.api response, current heartbeat timestamp)
    """

    root_logger.debug(f'send_heartbeat({nb_instance}, {nb_telemetry}, {nb_status_id},'
                      f' {previous_status_time})')

    status, _delete_attributes = nb_telemetry.diff(nb_telemetry.status_on_nuvla,
                                                   nb_telemetry.status)
    status_current_time = nb_telemetry.status.get('current-time', '')
    delete_attributes = []

    root_logger.debug(f'send_heartbeat: status_current_time = {status_current_time}  '
                      f'_delete_attributes = {_delete_attributes}  status = {status}')

    if not status_current_time:
        status = {'status-notes': ['NuvlaBox Telemetry is starting']}
        nb_telemetry.status.update(status)
    else:
        if status_current_time <= previous_status_time:
            status = {
                'status-notes': status.get('status-notes', []) + ['NuvlaBox telemetry is falling behind'],
                'status': 'DEGRADED'
            }
            nb_telemetry.status.update(status)
        else:
            delete_attributes = _delete_attributes

    # root_logger.debug('Refresh status: %s' % status)
    if delete_attributes:
        root_logger.info(f'Deleting the following attributes from NuvlaBox Status: '
                         f'{", ".join(delete_attributes)}')

    try:
        r = nb_instance.api().edit(nb_status_id,
                                   data=status,
                                   select=delete_attributes)
        nb_telemetry.status_on_nuvla.update(status)

    except:
        root_logger.error("Unable to update NuvlaBox status in Nuvla")
        raise

    return r.data, status_current_time


def wait_for_api_ready():
    """
    Waits in a loop for the API to be ready
    :return:
    """
    while True:
        try:
            r = requests.get('http://localhost/api/healthcheck')
            r.raise_for_status()
            if r.status_code == 200:
                break
        except:
            time.sleep(1)

    root_logger.info('NuvlaBox Agent has been initialized.')
    return


if __name__ == "__main__":
    socket.setdefaulttimeout(network_timeout)

    e = Event()
    # Try to activate the NuvlaBox
    activation = Activate(data_volume)
    root_logger.info(f'Nuvla endpoint: {activation.nuvla_endpoint}')
    root_logger.info(f'Nuvla connection insecure: {str(activation.nuvla_endpoint_insecure)}')
    while True:
        can_activate, user_info = activation.activation_is_possible()
        if can_activate or user_info:
            break

        e.wait(timeout=3)

    if not user_info:
        # this NuvlaBox hasn't been activated yet
        user_info = activation.activate()

    nuvlabox_resource, old_nuvlabox_resource = activation.update_nuvlabox_resource()
    nuvlabox_status_id = nuvlabox_resource["nuvlabox-status"]
    activation.vpn_commission_if_needed(nuvlabox_resource, old_nuvlabox_resource)

    telemetry = Telemetry(data_volume, nuvlabox_status_id,
                          os.environ.get('EXCLUDED_MONITORS', ''))
    infra = Infrastructure(data_volume)
    NB = NuvlaBoxCommon.NuvlaBoxCommon()

    if not infra.installation_home:
        root_logger.error('Host user HOME directory not defined. This might impact future'
                          ' SSH management actions')
    else:
        with open(infra.host_user_home_file, 'w') as userhome:
            userhome.write(infra.installation_home)
        infra.set_immutable_ssh_key()

    nuvlabox_info_updated_date = ''
    refresh_interval = 30

    app.config["telemetry"] = telemetry
    app.config["infra"] = infra

    api_thread = threading.Thread(target=app.run,
                                  kwargs={"host": "0.0.0.0", "port": "80"})
    api_thread.daemon = True
    api_thread.start()

    telemetry_thread = None

    can_continue = True
    nb_checker = None

    past_status_time = ''

    # start telemetry
    with NuvlaBoxCommon.timeout(10):
        root_logger.info('Waiting for API to be ready...')
        wait_for_api_ready()

    root_logger.info("Starting telemetry...")
    while True:
        start_cycle = time.time()

        if not can_continue:
            break

        if telemetry_thread and telemetry_thread.is_alive():
            root_logger.warning('Telemetry thread not completed in time')

        response, past_status_time = send_heartbeat(NB, telemetry,
                                                    nuvlabox_status_id, past_status_time)

        if not nb_checker or not nb_checker.is_alive():
            nb_checker = threading.Thread(target=preflight_check,
                                          args=(activation, infra,
                                                nuvlabox_info_updated_date,),
                                          daemon=True)
            nb_checker.start()

        if not telemetry_thread or not telemetry_thread.is_alive():
            telemetry_thread = threading.Thread(target=telemetry.update_status,
                                                daemon=True)
            telemetry_thread.start()

        if isinstance(response.get('jobs'), list) and \
                infra.container_runtime.job_engine_lite_image and \
                response.get('jobs'):

            root_logger.info(f'Processing the following jobs in pull-mode: '
                             f'{response["jobs"]}')

            threading.Thread(target=manage_pull_jobs,
                             args=(response['jobs'],
                                   infra.container_runtime.job_engine_lite_image,),
                             daemon=True).start()

        if not infra.is_alive():
            infra = Infrastructure(data_volume)
            infra.start()

        cycle_duration = time.time() - start_cycle
        next_cycle_in = refresh_interval - cycle_duration - 1
        root_logger.debug(f'End of cycle. Cycle duration: {cycle_duration} sec. Next '
                          f'cycle in {next_cycle_in} sec.')

        e.wait(timeout=next_cycle_in)
